temporal/temporalite/options.go

```diff
@@ -0,0 +1,166 @@
+// Unless explicitly stated otherwise all files in this repository are licensed under the MIT License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+
+package temporalite
```

Comment:
I’m a bit conflicted here. :duck: 

On one hand I 100% agree that we should avoid a proliferation of multiple “server flavor” packages. And wrapping multiple layers of option types between `temporaltest`, `temporalite`, and `temporal` packages is extremely awkward. _Some_ of the config options we have in `temporalite` might even make sense for other storage backends as well.

On the other hand, I think there is value in the current options chaining style of configuration we have in temporalite. The raison d’etre for this package was always to provide a high level interface for configuring a single node server, and IMO option functions make it easier to encapsulate complexity for a couple reasons:
* They can modify multiple pieces of config at once (eg. [WithDatabaseFilePath](https://github.com/temporalio/temporal/blob/d1ee6c6ed380e55000b309800abf9d92fbed1d2c/temporal/temporalite/options.go#L21-L27))
* They can be use case driven. This means users only need to care about "what" they want rather than "how" to do it. Eg. [WithSearchAttributeCacheDisabled](https://github.com/temporalio/temporal/blob/d1ee6c6ed380e55000b309800abf9d92fbed1d2c/temporal/temporalite/options.go#L119-L128) sets a dynamic config value under the hood, but the godoc comment for this function is only concerned with why you would want to do so. We're likely to have more cases like this for dynamic config and sqlite pragmas in the future.

I suppose a third option would be to implement option chaining on a config struct, eg.

```go
type TemporaliteConfig struct {
    DynamicConfig dynamicconfig.StaticClient
}

func (cfg *TemporaliteConfig) WithSearchAttributeCacheDisabled() *TemporaliteConfig {
    // modify dynamic config
}
```

And usage would look like

```go
temporal.NewTemporaliteServer(
    temporal.NewTemporaliteConfig()
        .WithSearchAttributeCacheDisabled(),
        .WithPersistenceDisabled(),
)
```

Reply:
I'm fine with the option chaining approach, we just need to have a different `New` call if we do use the `temporal` package which I'm also fine with (e.g. `NewSQLiteServer` or something).

Also, a separate package may make sense since we're gonna have to have a stronger API commitment on it.

Not the biggest fan of the term "temporalite" compared to all the other normal terms we name packages.

But regardless, I don't have a strong opinions here. Need server feedback. cc @dnr

Reply:
After further thought I would prefer to stick with the current approach of separate packages and usage of option list args. This should help us keep the `temporalite` package focused on the mission of simple, opinionated, and stable configuration of single-node Temporal servers, while the `temporal` package caters to anyone who needs absolute customization.

If we agree on that, then I think the remaining issue is to find a package name we can all accept. Here's my short list, and I'm very open to other suggestions:

* `temporalite` -- My top preference; it's a familiar callback to the original Temporalite as we sunset that project and also to SQLite (without necessarily tying us to a SQLite implementation forever).
* `temporalembedded` or `temporal/embedded` -- My only concern is that this could imply that the package might not be useful for a "standalone" server mode, though in that case you could probably use `temporalio/cli` instead.
* `temporalsinglenode` or `temporal/singlenode` -- This could be confused with running all Temporal services in a single node, but with an external database. That's an interesting use case, but I think better suited for options within the existing `temporal` package.
* `temporal/local` -- Local dev/test servers are a top use case for Temporalite, but I would like to avoid bias against deploying Temporalite instances to servers & other production usage in the long term.

Reply:
[I started writing this before @jlegrone's latest comment]

Lots of options here (no pun intended)...

Fyi, I don't know much about this history of this package/code so apologies if I say something controversial.

I also don't like the idea of multiple option types for different flavors of server. We already have ServerOptions which is designed to be extensible. I'd love to see if we can consolidate things into a single entry point, or possibly a single option type with two entry points. That might mean the current option type + validation gets somewhat more complicated (i.e. setting a dynamic config client would conflict with explicit dynamic config values).

(I'm not a fan of the chaining approach, but it works just as well and is more readable to me if you just pass a list of options instead of chaining call.)

As for the actual options, I'm also seeing some overlap between these options and the static config file. I wonder how people feel about the idea of putting everything in a config struct that can be mostly loaded from yaml or statically defined in code, with some fields being interfaces.

Reply:
Thanks for the feedback. I've refactored the `temporalite` package into the main `temporal` package, and added a `NewLiteServer` func. Let me know how that looks!